pkg/storegateway/series_refs_test.go

```diff
@@ -1630,6 +1630,124 @@ func TestOpenBlockSeriesChunkRefsSetsIterator(t *testing.T) {
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_2"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_3"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_4"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_5"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_6"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_7"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_8"+labelLongSuffix, "s", "foo")},
+				}}, {series: []seriesChunkRefs{
+					{lset: labels.FromStrings("i", "100"+labelLongSuffix, "j", "bar", "n", "1_9"+labelLongSuffix, "s", "foo")},
+				}},
+			},
+		},
+		"applies pending matchers when they match no series": {
+			matchers:        []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", ""), labels.MustNewMatcher(labels.MatchEqual, "s", "foo"), labels.MustNewMatcher(labels.MatchRegexp, "i", "100.*")},
+			pendingMatchers: []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", "")},
+			batchSize:       100,
+			expectedSeries:  []seriesChunkRefsSet{},
+		},
+	}
+
+	for testName, testCase := range testCases {
+		testName, testCase := testName, testCase
+		t.Run(testName, func(t *testing.T) {
+			require.Subset(t, testCase.matchers, testCase.pendingMatchers, "pending matchers should be a subset of all matchers")
+
+			var block = newTestBlock()
+			block.pendingReaders.Add(1) // this is hacky, but can be replaced only block.indexReade() accepts a strategy
+			indexReader := newBucketIndexReader(block, omitMatchersStrategy{testCase.pendingMatchers})
+			defer indexReader.Close()
+
+			hashCache := hashcache.NewSeriesHashCache(1024 * 1024).GetBlockCache(block.meta.ULID.String())
+
+			minT, maxT := block.meta.MinTime, block.meta.MaxTime
+			iterator, err := openBlockSeriesChunkRefsSetsIterator(
+				ctx,
+				testCase.batchSize,
+				"",
+				indexReader,
+				newInMemoryIndexCache(t),
+				block.meta,
+				testCase.matchers,
+				nil,
+				cachedSeriesHasher{hashCache},
+				true, // skip chunks since we are testing labels filtering
+				minT,
+				maxT,
+				2,
+				newSafeQueryStats(),
+				nil,
+			)
+			require.NoError(t, err)
+
+			actualSeriesSets := readAllSeriesChunkRefsSet(iterator)
+			assertSeriesChunkRefsSetsEqual(t, block.meta.ULID, testCase.expectedSeries, actualSeriesSets)
```

Comment:
[nit] Instead of defining the expected series in the test case, have you considered running `openBlockSeriesChunkRefsSetsIterator()` twice: one with `omitMatchersStrategy` (actual) and one without (expected). We expect `readAllSeriesChunkRefsSet()` returns the same set of series for both of them.

Reply:
i like this. I added it in c6bbc3ed02dcda26c07e4c3ce6f75e5333c6a504